### PR TITLE
[dom-gpu] Add HPX CUDA recipes and update production lists

### PR DESCRIPTION
**Pull request by msimberg**

### CODE REVIEW

**Comment by lucamar** [comment]
retest this please

**Comment by msimberg** [comment]
Could this be retriggered, please? The builds seem to have failed but it looks like the logs have already been cleaned up.

**Comment by lucamar** [comment]
retest this please

**Comment by lucamar** [comment]
@msimberg: I can see the following error in the log
```
ERROR: Failed to process easyconfig /scratch/snx3000/jenscscs/daint101/workspace/TestingEB/easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1-APEX.eb: Parsing easyconfig file failed: name 'cudaversion' is not defined (line 6)
```

**Comment by msimberg** [comment]
That was indeed from my commit this morning. Thanks for fixing it!

**Comment by lucamar** [comment]
The build of `HPX-1.4.0-CrayGNU-19.10-cuda-10.1-APEX.eb` failed after `[ 76%] Built target cublas_matmul`. I can see one warning in the log: 
```
WARNING Found 1 errors in command output (output: -- Performing Test HPX_WITH_CXX17_NOEXCEPT_FUNCTIONS_AS_NONTYPE_TEMPLATE_ARGUMENTS - Failed)
```
See `/scratch/snx3000/jenscscs/TestingEB/daint-gpu/40eaa6f-3112/tmp/eb-j3PYWW/easybuild-HPX-1.4.0-20200210.104408.weeMZ.log`

**Comment by msimberg** [comment]
That warning isn't a problem but there are some other build errors. I'll try building it manually.

**Comment by msimberg** [comment]
Uhm, how do I reproduce the exact same environment that jenkins uses? The 1.4.0 release builds fine in my environments also with CUDA enabled.

**Comment by lucamar** [comment]
@msimberg Jenkins build on a compute node: have you tried it as well?

**Comment by msimberg** [comment]
All right, I managed to reproduce the problem now and for some reason it seems like the `-craype-verbose` flag is causing the build problems. I can run two otherwise identical builds with the only difference being the existence or non-existence of that flag and the build with that flag fails. Is there a way to turn that off?

File: easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1-APEX.eb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+homepage = 'http://stellar-group.org/libraries/hpx'
+description = """HPX (High Performance ParalleX) is a general purpose C++
+runtime system for parallel and distributed applications of any scale. It
+strives to provide a unified programming model which transparently utilizes
+the available resources to achieve unprecedented levels of scalability.
+"""
+
+toolchain = {'name': 'CrayGNU', 'version': '19.10'}
+toolchainopts = {'pic': True}
```
Comment (lucamar) [inline easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1-APEX.eb]:
```suggestion
toolchainopts = {'pic': True, 'verbose': False}
```

Comment (lucamar) [reply easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1-APEX.eb]: @msimberg please try `'verbose': False` as I suggest

File: easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1-nonetworking.eb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+homepage = 'http://stellar-group.org/libraries/hpx'
+description = """HPX (High Performance ParalleX) is a general purpose C++
+runtime system for parallel and distributed applications of any scale. It
+strives to provide a unified programming model which transparently utilizes
+the available resources to achieve unprecedented levels of scalability.
+"""
+
+toolchain = {'name': 'CrayGNU', 'version': '19.10'}
+toolchainopts = {'pic': True}
```
Comment (lucamar) [inline easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1-nonetworking.eb]:
```suggestion
toolchainopts = {'pic': True, 'verbose': False}
```

File: easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1.eb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+homepage = 'http://stellar-group.org/libraries/hpx'
+description = """HPX (High Performance ParalleX) is a general purpose C++
+runtime system for parallel and distributed applications of any scale. It
+strives to provide a unified programming model which transparently utilizes
+the available resources to achieve unprecedented levels of scalability.
+"""
+
+toolchain = {'name': 'CrayGNU', 'version': '19.10'}
+toolchainopts = {'pic': True}
```
Comment (lucamar) [inline easybuild/easyconfigs/h/HPX/HPX-1.4.0-CrayGNU-19.10-cuda-10.1.eb]:
```suggestion
toolchainopts = {'pic': True, 'verbose': False}
```

**Comment by lucamar** [comment]
@msimberg Please try the change that I have suggested: as soon as you commit it, the build will be re-triggered

**Comment by msimberg** [comment]
How should I interpret the current failure?

**Comment by gppezzi** [comment]
retest this please

**Comment by gppezzi** [comment]
> How should I interpret the current failure?

Dom just came back from maintenance, re-trying to check if it was just a glitch...

**Comment by msimberg** [comment]
Just to follow up on this, is the behaviour with `-craype-verbose` something you've seen with other packages as well? It seems like that flag should never affect the behaviour of compilation, i.e. it seems like a bug to me...

**Comment by lucamar** [comment]
@msimberg I'm not sure: however, there is still some issue with the recipe, since the build failed on Dom: see https://jenkins.cscs.ch/blue/organizations/jenkins/ProductionEB/detail/ProductionEB/1017/pipeline

**Comment by gppezzi** [comment]
@msimberg We had similar issues with `-craype-verbose` in the past, typically this was caused by builds that at some point call the compilers directly instead of using the cray wrappers.

**Comment by msimberg** [comment]
```
== 2020-02-14 16:34:07,706 build_log.py:169 ERROR EasyBuild crashed with an error (at easybuild/software/EasyBuild/4.1.1/lib/python2.7/site-packages/easybuild/base/exceptions.py:124 in __init__): Sanity check failed: loading fake module failed: "PrgEnv-gnu/6.0.6(234):ERROR:105: Unable to locate a modulefile for 'udreg'" (at easybuild/software/EasyBuild/4.1.1/lib/python2.7/site-packages/easybuild/framework/easyblock.py:2596 in _sanity_check_step)
== 2020-02-14 16:34:07,706 easyblock.py:3109 WARNING build failed (first 300 chars): Sanity check failed: loading fake module failed: "PrgEnv-gnu/6.0.6(234):ERROR:105: Unable to locate a modulefile for 'udreg'"
```
I don't see anything caused by the HPX recipe in the log file. Am I missing something?

**Comment by lucamar** [comment]
@msimberg The issue was in the programming environment and not in the recipe: now the system administrators have restored the old one and we have triggered a new build of the software stack https://jenkins.cscs.ch/blue/organizations/jenkins/ProductionEB/detail/ProductionEB/1019/pipeline/35

**Pull request merged**